### ISSUE TITLE
MIR enabled rustc nightly warns about error in unreachable code

### ISSUE THREAD

**Issue by nwin** [rust-lang/rust#35316]
Compiling the following code using rustc+mir

``` rust
fn main() {
    let a = [1];
    if 1 == 1 {
        println!("{}", a[0])
    } else {
        println!("{}", a[1])
    }
}
```

emits the following warning

```
warning: this expression will panic at run-time
 --> <anon>:6:24
  |
6 |         println!("{}", a[1])
  |                        ^^^^ index out of bounds: the len is 1 but the index is 1
<std macros>:2:27: 2:58 note: in this expansion of format_args!
<std macros>:3:1: 3:54 note: in this expansion of print! (defined in <std macros>)
<anon>:6:9: 6:29 note: in this expansion of println! (defined in <std macros>)
```

although the `else` branch is unreachable. Switching off MIR by using `-Z orbit=off` causes the warning to disappear.

Edit: It seems that `rustc` lost the ability to detect this unreachable branch. Using `if true {` instead causes the warning to disappear as well.

**Comment by nagisa** [comment]
Constant propagation should help with this.

**Comment by eddyb** [comment]
@nwin rustc never actually checked, this warning is a new MIR feature.
We do need constant propagation to make it not trigger in cases like this.

**Comment by Mark-Simulacrum** [comment]
This is still the case today; however, I'm not sure that there's much purpose in keeping this issue around since it doesn't track anything specific other than "errors in unreachable code need to not happen," which as far as I can tell is really only fixable with constant propagation -- which this doesn't track.

**Comment by Mark-Simulacrum** [comment]
I'm going to close -- there's nothing inherently wrong with warning about unreachable code, and this issue isn't going to influence constant propagation being introduced into MIR/rustc.
